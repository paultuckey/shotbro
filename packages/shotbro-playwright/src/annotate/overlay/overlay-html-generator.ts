import {renderText} from "../shape/shape-text";
import {renderBox} from "../shape/shape-box";
import {renderCircle} from "../shape/shape-circle";
import {ShotBroInput} from "../annotate-types";
import {toStyleAttr} from "../shape/css-util";
import {InputPositions} from "../../main-shot/main-screenshotter";

export async function generateHtmlForOverlayString(input: ShotBroInput, inputPositions: InputPositions) {
  console.log('generate html for overlay');
  const bodyStyle = {
    width: `${inputPositions.focusBoxPosition!.w}px`,
    height: `${inputPositions.focusBoxPosition!.h}px`,
    overflow: 'hidden'
  }

  const shapesHtml = [];
  if (input.shapes) {
    for (let i = 0; i < input.shapes.length; i++) {
      let shape = input.shapes[i];
      let shapePos = inputPositions.shapePositions[i];
      if (shape && shapePos) {
        let html: string = '';
        if (shape.box) html = await renderBox(shapePos, shape.box);
        if (shape.circle) html = await renderCircle(shapePos, shape.circle);
        if (shape.text) html = await renderText(shapePos, shape.text);
        shapesHtml.push(html);
      }
    }
  }
  /**
   * we need to ensure there are minimal cross OS differences or else matching basline images won't work
   * use a standard font
   */
  const template = `
      <html lang="en">
      <head>
        <title>Overlay</title>
        <link href="https://cdn.jsdelivr.net/npm/modern-normalize@1.1.0/modern-normalize.css" rel="stylesheet"/>
        <link href="https://cdn.jsdelivr.net/npm/@fontsource/inter@4.5.12/500.css" rel="stylesheet"/>
        <style>
          body, h1 {
            font-size: 80px;
            font-family: Inter, sans-serif;
            background-color: transparent;
          }
        </style>
      </head>
      <body style="${toStyleAttr(bodyStyle)}">
        ${shapesHtml.join('')}
      </body>
      </html>
    `
  return template;
}


