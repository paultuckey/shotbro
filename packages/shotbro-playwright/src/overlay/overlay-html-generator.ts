import {renderBox} from "../shape/shape-box";
import {renderCircle} from "../shape/shape-circle";
import {type ShotBroBox, ShotBroInput} from "../annotate/annotate-types";
import {InputPositions} from "../main-shot/main-screenshot";
import {CliLog} from "../util/log";
import {renderIcon} from "../shape/shape-icon";

import {n2dp} from "../shape/shape-utils";

/**
 * Amount of spare space to the left and top of the main screenshot.  Only apply at the very last second when making CSS.
 */
export const BLEED = 100;

export async function generateHtmlForOverlayString(
    log: CliLog,
    mainPng: string,
    input: ShotBroInput,
    inputPositions: InputPositions,
    debug: boolean | undefined,
    bundledAssetsUrl: string | undefined
) {
    const shapesHtml = [];
    log.debug("generateHtmlForOverlayString", mainPng, debug, bundledAssetsUrl);
    if (input.shapes) {
        for (let i = 0; i < input.shapes.length; i++) {
            let shape = input.shapes[i];
            let shapeUniqId = 'shape' + i;
            let shapePos = inputPositions.shapePositions[i];
            if (shape && shapePos) {
                let html: string = '';
                if (shape.box) html = await renderBox(shapeUniqId, shapePos, shape);
                if (shape.circle) html = await renderCircle(shapeUniqId, shapePos, shape);
                if (shape.icon) html = await renderIcon(shapeUniqId, shapePos, shape);
                shapesHtml.push(html);
                if (debug) {
                    let posHelper = debug ? renderHelperBox(shapeUniqId, shapePos) : '';
                    shapesHtml.push(posHelper);
                }
            }
        }
    }
    if (inputPositions.focusBoxPosition && debug) {
        let posHelper = debug ? renderHelperBox('focus', inputPositions.focusBoxPosition) : '';
        shapesHtml.push(posHelper);
    }

    /**
     * Note use of normalize and open sans font is to ensure there are minimal cross OS differences.
     */
    const template = `
      <!DOCTYPE html>
      <html lang="en">
      <head>
        <title>Overlay</title>
        <link href="${bundledAssetsUrl}/@fontsource/open-sans/500.css" rel="stylesheet"/>
        <link href="${bundledAssetsUrl}/bootstrap/dist/css/bootstrap.css" rel="stylesheet"/>
        <link href="${bundledAssetsUrl}/bootstrap-icons/font/bootstrap-icons.css" rel="stylesheet"/>
        <style>
          body {
            font-family: "Open Sans", sans-serif;
            background-color: transparent;
            background-image: url('${mainPng}');
            background-position: ${BLEED}px ${BLEED}px;
            background-repeat: no-repeat;
          }
        </style>
      </head>
      <body>
      ${shapesHtml.join('\n\n')}
      </body>
      </html>`
    return template;
}

function renderHelperBox(scope: string, elPos: ShotBroBox): string {
    return `
        <style>
        .${scope}.helper {
            position: absolute;
            top: ${n2dp(elPos.y + BLEED)}px;
            left: ${n2dp(elPos.x + BLEED)}px;
            width: ${n2dp(elPos.w)}px;
            height: ${n2dp(elPos.h)}px;
            border: 1px solid red;
          }
         </style>
        <div id="helper-${scope}" class="${scope} helper"></div>`;
}
