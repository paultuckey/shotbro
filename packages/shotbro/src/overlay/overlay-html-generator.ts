import {renderText} from "../shape/shape-text";
import {renderBox} from "../shape/shape-box";
import {renderCircle} from "../shape/shape-circle";
import {ShapeBoxes} from "../shape/shape-locator";
import {ShotBroBox, ShotBroInput} from "../shotbro-types";
import {ShapeCommon} from "../shape/shape-types";
import {toStyleAttr} from "../shape/css-util";

export async function generateHtmlForOverlayString(shotBroProps: ShotBroInput, mainScreenShotBox: ShotBroBox,
                                                   foundElementBoxes: ShapeBoxes) {
  console.log('generate html for overlay');
  let getElPos = (shape: ShapeCommon) : ShotBroBox => {
    let foundBox = {w: 0, h: 0, x: 0, y: 0 }
    if (shape?.id) {
      foundBox = foundElementBoxes[shape.id]
    } else {
      console.log('missing elPos for shape', shape.id)
    }
    return foundBox;
  }

  const bodyStyle = {
    width: `${mainScreenShotBox.w}px`,
    height: `${mainScreenShotBox.h}px`,
    overflow: 'hidden'
  }

  const shapesHtml = [];
  if (shotBroProps.shapes) {
    for (let shape of shotBroProps.shapes) {
      let html: string = '';
      if (shape?.box) html = await renderBox(getElPos(shape?.box), shape?.box);
      if (shape?.circle) html = await renderCircle(getElPos(shape?.circle), shape?.circle);
      if (shape?.text) html = await renderText(getElPos(shape?.text), shape?.text);
      shapesHtml.push(html);
    }
  }
  /**
   * we need to ensure there are minimal cross OS differences or else matching basline images won't work
   * use a standard font
   */
  const template = `
      <html>
      <head>
        <title>Overlay</title>
        <link href="https://cdn.jsdelivr.net/npm/modern-normalize@1.1.0/modern-normalize.css" rel="stylesheet"/>
        <link href="https://cdn.jsdelivr.net/npm/@fontsource/inter@4.5.12/500.css" rel="stylesheet"/>
        <style>
          body, h1 {
            font-size: 80px;
            font-family: Inter, sans-serif;
            background-color: transparent;
          }
        </style>
      </head>
      <body style="${toStyleAttr(bodyStyle)}">
        ${shapesHtml.join('')}
      </body>
      </html>
    `
  return template;
}


