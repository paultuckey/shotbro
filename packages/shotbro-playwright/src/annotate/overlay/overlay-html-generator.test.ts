import {generateHtmlForOverlayString} from "./overlay-html-generator";
import type {ShotBroInput} from "../annotate-types";

describe('Shapes HTML Generator', () => {

    test('Simple test', async () => {
        const input: ShotBroInput = {
            shotName: 'test',
            shapes: [
                {circle: {id: 'circle1'}}
            ]
        }

        const html = await generateHtmlForOverlayString(input,
            {
                focusBoxPosition: {x: 0, y: 0, w: 100, h: 100},
                shapePositions: [{x: 4, y: 5, w: 40, h: 50}]
            });
        expect(html).toMatchSnapshot('shapes-simple');
    })

    test('Style kitchen sink', async () => {
        const input: ShotBroInput = {
            shotName: 'test',
            shapes: [
                {circle: {id: 'circle1'}}, {circle: {id: 'circle2'}},
                {box: {id: 'box1'}}, {box: {id: 'box2'}},
                {text: {id: 'text1', value: 'Text1'}}, {text: {id: 'text2', value: 'Text2'}},
            ]
        }

        const html = await generateHtmlForOverlayString(input, {
            focusBoxPosition: {x: 0, y: 0, w: 400, h: 400},
            shapePositions: [
                {x: 20, y: 100, w: 100, h: 100},
                {x: 200, y: 100, w: 100, h: 50},
                {x: 20, y: 300, w: 100, h: 100},
                {x: 200, y: 300, w: 100, h: 50},
                {x: 20, y: 500, w: 100, h: 100},
                {x: 200, y: 500, w: 100, h: 50},
            ]
        });
        //fs.writeFileSync(path.join(__dirname, SNAPSHOTS_DIR_NAME, 'style.html'), html, 'utf-8')
        expect(html).toMatchSnapshot('shapes-style');
    })

});