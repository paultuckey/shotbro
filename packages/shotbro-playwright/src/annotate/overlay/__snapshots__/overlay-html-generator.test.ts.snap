// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shapes HTML Generator Simple test: shapes-simple 1`] = `
"
      <html>
      <head>
        <title>Overlay</title>
        <link href="https://cdn.jsdelivr.net/npm/modern-normalize@1.1.0/modern-normalize.css" rel="stylesheet"/>
        <link href="https://cdn.jsdelivr.net/npm/@fontsource/inter@4.5.12/500.css" rel="stylesheet"/>
        <style>
          body, h1 {
            font-size: 80px;
            font-family: Inter, sans-serif;
            background-color: transparent;
          }
        </style>
      </head>
      <body style="width:100px;height:100px;overflow:hidden;">
        <div style="position:fixed;top:3px;left:2px;width:40px;height:50px;border-width:4px;border-style:solid;border-color:deeppink;border-radius:20px / 25px;box-shadow:2px 2px 2px 0px rgba(0,0,0,0.13);"></div>
      </body>
      </html>
    "
`;

exports[`Shapes HTML Generator Style kitchen sink: shapes-style 1`] = `
"
      <html>
      <head>
        <title>Overlay</title>
        <link href="https://cdn.jsdelivr.net/npm/modern-normalize@1.1.0/modern-normalize.css" rel="stylesheet"/>
        <link href="https://cdn.jsdelivr.net/npm/@fontsource/inter@4.5.12/500.css" rel="stylesheet"/>
        <style>
          body, h1 {
            font-size: 80px;
            font-family: Inter, sans-serif;
            background-color: transparent;
          }
        </style>
      </head>
      <body style="width:320px;height:500px;overflow:hidden;">
        <div style="position:fixed;top:98px;left:18px;width:100px;height:100px;border-width:4px;border-style:solid;border-color:deeppink;border-radius:50px / 50px;box-shadow:2px 2px 2px 0px rgba(0,0,0,0.13);"></div><div style="position:fixed;top:98px;left:198px;width:100px;height:50px;border-width:4px;border-style:solid;border-color:deeppink;border-radius:50px / 25px;box-shadow:2px 2px 2px 0px rgba(0,0,0,0.13);"></div><div style="position:fixed;top:298px;left:18px;width:108px;height:108px;border-width:4px;border-style:solid;border-color:deeppink;border-radius:4px;box-shadow:2px 2px 2px 0px rgba(0,0,0,0.13);"></div><div style="position:fixed;top:298px;left:198px;width:108px;height:58px;border-width:4px;border-style:solid;border-color:deeppink;border-radius:4px;box-shadow:2px 2px 2px 0px rgba(0,0,0,0.13);"></div><div style="position:fixed;top:500px;left:20px;width:100px;height:100px;font-size:32px;color:deeppink;text-shadow:2px 2px rgba(0,0,0,0.13);">Text1</div><div style="position:fixed;top:500px;left:200px;width:100px;height:50px;font-size:32px;color:deeppink;text-shadow:2px 2px rgba(0,0,0,0.13);">Text2</div>
      </body>
      </html>
    "
`;
